src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -472,9 +487,18 @@ void ObjectSynchronizer::enter(Handle obj, BasicLock* lock, JavaThread* current)
     if (LockingMode == LM_LIGHTWEIGHT) {
       // Fast-locking does not use the 'lock' argument.
       LockStack& lock_stack = current->lock_stack();
-      if (lock_stack.can_push()) {
-        markWord mark = obj()->mark_acquire();
-        while (mark.is_neutral()) {
+      if (lock_stack.is_full()) {
+        // The emitted code always goes into the runtime in case the lock stack
+        // is full. We unconditionally make room on the lock stack by inflating
+        // the least recently locked object on the lock stack.
+        ObjectMonitor* monitor = inflate(current, lock_stack.bottom(), inflate_cause_vm_internal);
+        assert(monitor->owner() == current, "must be owner=" PTR_FORMAT " current=" PTR_FORMAT " mark=" PTR_FORMAT,
+               p2i(monitor->owner()), p2i(current), monitor->object()->mark_acquire().value());
+        assert(!lock_stack.is_full(), "must have made room here");
+      }
```

Comment:
This is an interesting idea. I'll have to see how you test this code later on...

Reply:
Sure. They are generally tricky, you usually need to tell the compiler not to do stuff like eliding locks, and inline methods. I will attempt to write tests both for -Xint and the compiler that tests both recursive and normal full lock stack. (The recursive tests requires the platform to implement recursive lightweight for it to effectively test anything)

Reply:
I wrote a little test case just now.  Maybe deciding which object should cause inflation can be tuned with some benchmarking results if we find some degenerate case?  Maybe the top lock should be inflated since that might be the one that causes the stack to overflow?

Reply:
In any case, maybe there should be a logging statement here for this.